File: packages/code-storage-typescript/src/commit-pack.ts
import { RefUpdateError, inferRefUpdateReason } from './errors';
import type { CommitPackAckRaw } from './schemas';
import { commitPackResponseSchema, errorEnvelopeSchema } from './schemas';
import type { CommitResult, RefUpdate } from './types';

export type CommitPackAck = CommitPackAckRaw;

export function buildCommitResult(ack: CommitPackAckRaw): CommitResult {
  const refUpdate = toRefUpdate(ack.result);
  if (!ack.result.success) {
    throw new RefUpdateError(
      ack.result.message ?? `Commit failed with status ${ack.result.status}`,
      {
        status: ack.result.status,
        message: ack.result.message,
        refUpdate,
      }
    );
  }
  return {
    commitSha: ack.commit.commit_sha,
    treeSha: ack.commit.tree_sha,
    targetBranch: ack.commit.target_branch,
    packBytes: ack.commit.pack_bytes,
    blobCount: ack.commit.blob_count,
    refUpdate,
  };
}

export function toRefUpdate(result: CommitPackAckRaw['result']): RefUpdate {
  const targetBranch = result.target_branch ?? result.branch ?? '';
  return {
    targetBranch,
    branch: targetBranch,
    oldSha: result.old_sha,
    newSha: result.new_sha,
  };
}

export async function parseCommitPackError(
  response: Response,
  fallbackMessage: string
): Promise<{
  statusMessage: string;
  statusLabel: string;
  refUpdate?: Partial<RefUpdate>;
}> {
  const cloned = response.clone();
  let jsonBody: unknown;
  try {
    jsonBody = await cloned.json();
  } catch {
    jsonBody = undefined;
  }

  let textBody: string | undefined;
  if (jsonBody === undefined) {
    try {
      textBody = await response.text();
    } catch {
      textBody = undefined;
    }
  }

  const defaultStatus = (() => {
    const inferred = inferRefUpdateReason(String(response.status));
    return inferred === 'unknown' ? 'failed' : inferred;
  })();
  let statusLabel = defaultStatus;
  let refUpdate: Partial<RefUpdate> | undefined;
  let message: string | undefined;

  if (jsonBody !== undefined) {
    const parsedResponse = commitPackResponseSchema.safeParse(jsonBody);
    if (parsedResponse.success) {
      const result = parsedResponse.data.result;
      if (typeof result.status === 'string' && result.status.trim() !== '') {
        statusLabel = result.status.trim() as typeof statusLabel;
      }
      refUpdate = toPartialRefUpdateFields(
        result.target_branch ?? result.branch,
        result.old_sha,
        result.new_sha
      );
      if (typeof result.message === 'string' && result.message.trim() !== '') {
        message = result.message.trim();
      }
    }

    if (!message) {
      const parsedError = errorEnvelopeSchema.safeParse(jsonBody);
      if (parsedError.success) {
        const trimmed = parsedError.data.error.trim();
        if (trimmed) {
          message = trimmed;
        }
      }
    }
  }

  if (!message && typeof jsonBody === 'string' && jsonBody.trim() !== '') {
    message = jsonBody.trim();
  }

  if (!message && textBody && textBody.trim() !== '') {
    message = textBody.trim();
  }

  return {
    statusMessage: message ?? fallbackMessage,
    statusLabel,
    refUpdate,
  };
}

function toPartialRefUpdateFields(
  branch?: string | null,
  oldSha?: string | null,
  newSha?: string | null
): Partial<RefUpdate> | undefined {
  const refUpdate: Partial<RefUpdate> = {};

  if (typeof branch === 'string' && branch.trim() !== '') {
    refUpdate.targetBranch = branch.trim();
    refUpdate.branch = branch.trim();
  }
  if (typeof oldSha === 'string' && oldSha.trim() !== '') {
    refUpdate.oldSha = oldSha.trim();
  }
  if (typeof newSha === 'string' && newSha.trim() !== '') {
    refUpdate.newSha = newSha.trim();
  }

  return Object.keys(refUpdate).length > 0 ? refUpdate : undefined;
}

File: packages/code-storage-go/commit_pack.go
package storage

import (
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strconv"
	"strings"
)

type commitPackAck struct {
	Commit struct {
		CommitSHA    string `json:"commit_sha"`
		TreeSHA      string `json:"tree_sha"`
		TargetBranch string `json:"target_branch"`
		PackBytes    int    `json:"pack_bytes"`
		BlobCount    int    `json:"blob_count"`
	} `json:"commit"`
	Result struct {
		TargetBranch *string `json:"target_branch"`
		Branch       string  `json:"branch"`
		OldSHA       string  `json:"old_sha"`
		NewSHA       string  `json:"new_sha"`
		Success      bool    `json:"success"`
		Status       string  `json:"status"`
		Message      string  `json:"message,omitempty"`
	} `json:"result"`
}

type commitPackResponse struct {
	Commit *struct {
		CommitSHA    string `json:"commit_sha"`
		TreeSHA      string `json:"tree_sha"`
		TargetBranch string `json:"target_branch"`
		PackBytes    int    `json:"pack_bytes"`
		BlobCount    int    `json:"blob_count"`
	} `json:"commit,omitempty"`
	Result struct {
		TargetBranch *string `json:"target_branch"`
		Branch       string  `json:"branch"`
		OldSHA       string  `json:"old_sha"`
		NewSHA       string  `json:"new_sha"`
		Success      *bool   `json:"success"`
		Status       string  `json:"status"`
		Message      string  `json:"message"`
	} `json:"result"`
}

type errorEnvelope struct {
	Error string `json:"error"`
}

func buildCommitResult(ack commitPackAck) (CommitResult, error) {
	targetBranch := preferredResponseString(ack.Result.TargetBranch, ack.Result.Branch)
	refUpdate := RefUpdate{
		TargetBranch: targetBranch,
		Branch:       targetBranch,
		OldSHA:       ack.Result.OldSHA,
		NewSHA:       ack.Result.NewSHA,
	}

	if !ack.Result.Success {
		message := ack.Result.Message
		if strings.TrimSpace(message) == "" {
			message = "commit failed with status " + ack.Result.Status
		}
		return CommitResult{}, newRefUpdateError(message, ack.Result.Status, &refUpdate)
	}

	return CommitResult{
		CommitSHA:    ack.Commit.CommitSHA,
		TreeSHA:      ack.Commit.TreeSHA,
		TargetBranch: ack.Commit.TargetBranch,
		PackBytes:    ack.Commit.PackBytes,
		BlobCount:    ack.Commit.BlobCount,
		RefUpdate:    refUpdate,
	}, nil
}

func parseCommitPackError(resp *http.Response, fallbackMessage string) (string, string, *RefUpdate, error) {
	body, err := readAll(resp)
	if err != nil {
		return "", "", nil, err
	}

	statusLabel := defaultStatusLabel(resp.StatusCode)
	var refUpdate *RefUpdate
	message := ""

	var parsed commitPackResponse
	if err := json.Unmarshal(body, &parsed); err == nil {
		if strings.TrimSpace(parsed.Result.Status) != "" {
			statusLabel = strings.TrimSpace(parsed.Result.Status)
		}
		if parsed.Result.Message != "" {
			message = strings.TrimSpace(parsed.Result.Message)
		}
		refUpdate = partialRefUpdate(preferredResponseString(parsed.Result.TargetBranch, parsed.Result.Branch), parsed.Result.OldSHA, parsed.Result.NewSHA)
	}

	if message == "" {
		var errEnv errorEnvelope
		if err := json.Unmarshal(body, &errEnv); err == nil {
			if strings.TrimSpace(errEnv.Error) != "" {
				message = strings.TrimSpace(errEnv.Error)
			}
		}
	}

	if message == "" && len(body) > 0 {
		message = strings.TrimSpace(string(body))
	}

	if message == "" {
		if fallbackMessage != "" {
			message = fallbackMessage
		} else {
			message = "commit request failed (" + strconv.Itoa(resp.StatusCode) + " " + resp.Status + ")"
		}
	}

	return message, statusLabel, refUpdate, nil
}

func defaultStatusLabel(statusCode int) string {
	status := inferRefUpdateReason(strconv.Itoa(statusCode))
	if status == RefUpdateReasonUnknown {
		return string(RefUpdateReasonFailed)
	}
	return string(status)
}

func partialRefUpdate(branch string, oldSHA string, newSHA string) *RefUpdate {
	branch = strings.TrimSpace(branch)
	oldSHA = strings.TrimSpace(oldSHA)
	newSHA = strings.TrimSpace(newSHA)

	if branch == "" && oldSHA == "" && newSHA == "" {
		return nil
	}
	return &RefUpdate{TargetBranch: branch, Branch: branch, OldSHA: oldSHA, NewSHA: newSHA}
}

func readAll(resp *http.Response) ([]byte, error) {
	if resp.Body == nil {
		return nil, errors.New("response body is empty")
	}
	return io.ReadAll(resp.Body)
}
